pkg/execution/executor/executor.go

```diff
@@ -551,22 +527,84 @@ func (e *executor) Execute(ctx context.Context, id state.Identifier, item queue.
 	// contains it.
 	md := s.Metadata()
 
+	// Retrieve the start time of the function
+	fnStartTime := time.Now()
+	if md.GetStartTime() != nil {
+		fnStartTime = *md.GetStartTime()
+	}
+
 	// Store the metadata in context for future use and propagate trace
 	// context. This can be used to reduce reads in the future.
 	ctx = e.extractTraceCtx(WithContextMetadata(ctx, md), id, &item)
 
-	ctx, span := telemetry.UserTracer().Provider().
-		Tracer(consts.OtelScopeExecution).
-		Start(ctx, "running", trace.WithAttributes(
+	// spanID should always exists
+	fnSpanID, err := md.GetSpanID()
+	if err != nil {
+		// generate a new one here to be used for subsequent runs.
+		// this could happen for runs that started before this feature was introduced.
+		sid := telemetry.NewSpanID(ctx)
+		fnSpanID = &sid
+	}
+
+	evtIDs := make([]string, len(id.EventIDs))
+	for i, eid := range id.EventIDs {
+		evtIDs[i] = eid.String()
+	}
+
+	sysDelay := time.Duration(md.SystemDelay)
+	sojDelay := time.Duration(md.SojournDelay)
+	if item.RunInfo != nil {
+		if md.SystemDelay == -1 {
+			sysDelay = item.RunInfo.Latency
+		}
+		if md.SojournDelay == -1 {
+			sojDelay = item.RunInfo.SojournDelay
+		}
+	}
+
+	// (re)Construct function span to force update the end time
+	ctx, fnSpan := telemetry.NewSpan(ctx,
+		telemetry.WithScope(consts.OtelScopeFunction),
+		telemetry.WithName(s.Function().GetSlug()),
+		telemetry.WithTimestamp(fnStartTime),
+		telemetry.WithSpanID(*fnSpanID),
+		telemetry.WithSpanAttributes(
 			attribute.Bool(consts.OtelUserTraceFilterKey, true),
 			attribute.String(consts.OtelSysAccountID, id.AccountID.String()),
 			attribute.String(consts.OtelSysWorkspaceID, id.WorkspaceID.String()),
 			attribute.String(consts.OtelSysAppID, id.AppID.String()),
 			attribute.String(consts.OtelSysFunctionID, id.WorkflowID.String()),
 			attribute.Int(consts.OtelSysFunctionVersion, id.WorkflowVersion),
 			attribute.String(consts.OtelAttrSDKRunID, id.RunID.String()),
-		))
-	defer span.End()
+			attribute.StringSlice(consts.OtelSysEventIDs, evtIDs),
+			attribute.String(consts.OtelSysIdempotencyKey, id.IdempotencyKey()),
+		),
+	)
+	if len(id.EventIDs) > 1 {
+		fnSpan.SetAttributes(attribute.String(consts.OtelSysBatchID, id.BatchID.String()))
+	}
+
+	ctx, span := telemetry.NewSpan(ctx,
+		telemetry.WithScope(consts.OtelScopeExecution),
+		telemetry.WithName("execute"),
```

Comment:
For steps that report (first parallel step and some following parallel steps), this span will linger. It sounded like the plan was to remove them entirely?

We likely need to call `span.Cancel()` further down the line for this, then?